### PR TITLE
UseHealthChecks vs. MapHealthChecks /5

### LINKED ISSUE #29421: Difference between UseHealthChecks and MapHealthChecks
This **Health checks in ASP.NET Core** page doesn't mention `UseHealthChecks` at all, and the few places in this repository that contain it, don't explain the difference between `UseHealthChecks` and `MapHealthChecks` very well. The [2.2 to 3.0 migration page](https://github.com/dotnet/AspNetCore.Docs/blob/57d21ee8b0701947b2d97dd0fa283a31c4a78e4d/aspnetcore/migration/22-to-30.md?plain=1#L1020) mentions benefits of authentication, but that wouldn't apply to the Kubernetes use that this page is using as an example.

Because this page mentions Kubernetes readiness and liveness probe, and these endpoints are called a lot, I think this distinction would especially helpful. I've seen it mention that one might want to use one vs. the other when using logging middleware of requests.

I view one of the advantages of `UseHealthChecks` is that it will terminate the pipeline if the path matches so it can be placed before the logging middleware and the logs will be skipped.

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: c0cbb094-29e9-740a-9fd6-db372641d753
* Version Independent ID: 52d8efc9-a336-71d9-6d90-98173cbf6bf4
* Content: [Health checks in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/host-and-deploy/health-checks?view=aspnetcore-7.0)
* Content Source: [aspnetcore/host-and-deploy/health-checks.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/host-and-deploy/health-checks.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-hostdeploy**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

---
[Associated WorkItem - 97769](https://dev.azure.com/msft-skilling/Content/_workitems/edit/97769)

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes #29421

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/host-and-deploy/health-checks.md](https://github.com/dotnet/AspNetCore.Docs/blob/c6ef09865ab6c35694cbcde34cc26841272f702b/aspnetcore/host-and-deploy/health-checks.md) | [Health checks in ASP.NET Core](https://review.learn.microsoft.com/en-us/aspnet/core/host-and-deploy/health-checks?branch=pr-en-us-29439) |

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
@JuergenGutsch please review. It's way too long, should I cut most of that out?

**Comment by Rick-Anderson** [comment]
@ngbrown please review

File: aspnetcore/host-and-deploy/health-checks.md [hunk 1]
```diff
@@ -300,10 +300,34 @@ The `SampleHealthCheckWithDiConfig` and the Health check needs to be added to th
 
 :::code language="csharp" source="~/host-and-deploy/health-checks/samples/7.x/HealthChecksSample/Snippets/Program.cs" id="snippet_MapHealthChecksUsingDependencyInjection":::
 
+## UseHealthChecks vs. MapHealthChecks
+
+The `MapHealthChecks` method accepts arguments similar to `UseHealthChecks`. The advantage of using `MapHealthChecks` over `UseHealthChecks` is the ability to apply authorization and to have greater fine-grained control over the matching policy. The primary advantage of using `UseHealthChecks` over `MapHealthChecks` over terminating the middleware and selecting endpoints via port.
+
+<xref:Microsoft.AspNetCore.Builder.HealthCheckApplicationBuilderExtensions.UseHealthChecks%2A>:
+* Terminates the pipeline when a request matches the health check endpoint. [Short-circuiting](xref:fundamentals/middleware/index) is often desirable because it avoids unnecessary work, such as logging and other middleware.
```
Comment (cremor) [inline aspnetcore/host-and-deploy/health-checks.md:308]: Is this difference actually by design? To be specific, is it by design that `MapHealthChecks` does **not** short-circuit?
Both methods register the same `HealthCheckMiddleware`. The only difference is how it is mapped. See [here](https://github.com/dotnet/aspnetcore/blob/d47e49e9c1e173ac90821f7e89cc38e710274241/src/Middleware/HealthChecks/src/Builder/HealthCheckEndpointRouteBuilderExtensions.cs#L77-L81) and [here](https://github.com/dotnet/aspnetcore/blob/d47e49e9c1e173ac90821f7e89cc38e710274241/src/Middleware/HealthChecks/src/Builder/HealthCheckApplicationBuilderExtensions.cs#L265).
Maybe related: https://github.com/aspnet/Diagnostics/pull/516

File: aspnetcore/host-and-deploy/health-checks.md [hunk 2]
```diff
@@ -300,10 +300,34 @@ The `SampleHealthCheckWithDiConfig` and the Health check needs to be added to th
...
+* Is primarily used for configuring the health check middleware in the pipeline.
+* Can select any endpoint on a port with a `null` or empty `PathString`, which can be useful for the following scenarios:
+  * Provides a default health check without specifying a path. Allows performing a health check on the root path or any request made to the specified port.
+  * Selecting any endpoint on a port.
+  * Provides a catch-all health check. Handles health checks for any path.
+* [Source code](https://github.com/dotnet/aspnetcore/blob/main/src/Middleware/HealthChecks/src/Builder/HealthCheckApplicationBuilderExtensions.cs)
+
+<xref:Microsoft.AspNetCore.Builder.HealthCheckEndpointRouteBuilderExtensions.MapHealthChecks%2A>:
+* Is used for mapping specific routes or endpoints for health checks.
+* Allows customization of the URL or path where the health check endpoint is accessible.  Additional options and parameters can be specified, such as customizing the HTTP status codes, response content, or applying filters to determine which health checks to execute.
```
Comment (cremor) [inline aspnetcore/host-and-deploy/health-checks.md:318]: All of that also seems possible with `UseHealthChecks` (via `HealthCheckOptions`).

**Comment by Rick-Anderson** [comment]
@CarlosLanderas @jods4 @mettelephant @FirstAfterGod2501 @bhugot please review

File: aspnetcore/host-and-deploy/health-checks.md (hunk 1 above)
Comment (ngbrown) [reply aspnetcore/host-and-deploy/health-checks.md:308]: The [routing documentation](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/routing) doesn't explain it very clearly, but the [.NET 5.0 to .NET 6.0 migration document](https://learn.microsoft.com/en-us/aspnet/core/migration/50-to-60-samples?view=aspnetcore-7.0#:~:text=Routes%20added%20directly%20to%20the%20WebApplication%20execute%20at%20the%20end%20of%20the%20pipeline.) says:

> Routes [i.e. `MapGet()`] added directly to the [WebApplication](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.builder.webapplication) execute at the _**end**_ of the pipeline.

**Comment by ngbrown** [comment]
@Rick-Anderson I think the updated explanation looks good.

**Comment by jods4** [comment]
@Rick-Anderson Not sure why I'm pinged on this doc change?
I commented on health checks before, it was to ask for a way to register health checks through DI. I don't think this change is related?

**Comment by mettelephant** [comment]
@Rick-Anderson I'm also not sure why I was pinged for this change either. I've opened an issue on health check logging before, but it doesn't seem to be related to this change.

**Comment by Rick-Anderson** [comment]
@JuergenGutsch please review. It's way too long, should I cut most of that out?

**Comment by cremor** [comment]
I seems like in .NET 8 we will be able to write `app.MapHealthChecks("/healthz").ShortCircuit();`, see #28935
I think this should also be mentioned here.

**Comment by Rick-Anderson** [comment]
> I seems like in .NET 8 we will be able to write `app.MapHealthChecks("/healthz").ShortCircuit();`, see #28935 I think this should also be mentioned here.

That was added in #29110

**Comment by Rick-Anderson** [comment]
> I seems like in .NET 8 we will be able to write `app.MapHealthChecks("/healthz").ShortCircuit();`, see #28935 I think this should also be mentioned here.

See #29709

File: aspnetcore/host-and-deploy/health-checks.md [hunk 3]
```diff
@@ -300,10 +300,34 @@ The `SampleHealthCheckWithDiConfig` and the Health check needs to be added to th
 
 :::code language="csharp" source="~/host-and-deploy/health-checks/samples/7.x/HealthChecksSample/Snippets/Program.cs" id="snippet_MapHealthChecksUsingDependencyInjection":::
 
+## UseHealthChecks vs. MapHealthChecks
+
+The `MapHealthChecks` method accepts arguments similar to `UseHealthChecks`. The advantage of using `MapHealthChecks` over `UseHealthChecks` is the ability to apply authorization and to have greater fine-grained control over the matching policy. The primary advantage of using `UseHealthChecks` over `MapHealthChecks` is terminating the middleware and selecting endpoints via port.
```
Comment (JamesNK) [inline aspnetcore/host-and-deploy/health-checks.md:305]: I think a summary of their implementation will help people understand their differences:

```suggestion
There are two ways to make health checks accessible to callers:

* <xref:Microsoft.AspNetCore.Builder.HealthCheckApplicationBuilderExtensions.UseHealthChecks%2A> registers middleware for handling health checks requests in the middleware pipeline.
* <xref:Microsoft.AspNetCore.Builder.HealthCheckEndpointRouteBuilderExtensions.MapHealthChecks%2A> registers a health checks endpoint. The endpoint is matched and executed along with other endpoints in the app.

The advantage of using `MapHealthChecks` over `UseHealthChecks` is the ability to use endpoint aware middleware, such as authorization, and to have greater fine-grained control over the matching policy. The primary advantage of using `UseHealthChecks` over `MapHealthChecks` is controlling exactly where health checks runs in the middleware pipeline.
```

File: aspnetcore/host-and-deploy/health-checks.md [hunk 4]
```diff
@@ -300,10 +300,34 @@ The `SampleHealthCheckWithDiConfig` and the Health check needs to be added to th
...
+
+<xref:Microsoft.AspNetCore.Builder.HealthCheckApplicationBuilderExtensions.UseHealthChecks%2A>:
+* Terminates the pipeline when a request matches the health check endpoint. [Short-circuiting](xref:fundamentals/middleware/index) is often desirable because it avoids unnecessary work, such as logging and other middleware.
+* Is primarily used for configuring the health check middleware in the pipeline.
+* Can select any endpoint on a port with a `null` or empty `PathString`, which can be useful for the following scenarios:
```
Comment (JamesNK) [inline aspnetcore/host-and-deploy/health-checks.md:310]: "Endpoint" has a meaning in ASP.NET Core of a route that matches to code. I don't think it makes sense here.

Is this what you mean:
```suggestion
* Can match any path on a port with a `null` or empty `PathString`, which can be useful for the following scenarios:
```

File: aspnetcore/host-and-deploy/health-checks.md [hunk 5]
```diff
@@ -300,10 +300,34 @@ The `SampleHealthCheckWithDiConfig` and the Health check needs to be added to th
...
+  * Provides a default health check without specifying a path. Allows performing a health check on the root path or any request made to the specified port.
+  * Selecting any endpoint on a port.
+  * Provides a catch-all health check. Handles health checks for any path.
```
Comment (JamesNK) [inline aspnetcore/host-and-deploy/health-checks.md:313]: These are all saying the same thing.

Comment (Rick-Anderson) [reply aspnetcore/host-and-deploy/health-checks.md:318]: > These are all saying the same thing.

@JamesNK  Can you make a suggestion of the lines to delete?

File: aspnetcore/host-and-deploy/health-checks.md [hunk 6]
```diff
@@ -300,10 +300,39 @@ The `SampleHealthCheckWithDiConfig` and the Health check needs to be added to th
...
+
+The advantage of using `MapHealthChecks` over `UseHealthChecks` is the ability to use endpoint aware middleware, such as authorization, and to have greater fine-grained control over the matching policy. The primary advantage of using `UseHealthChecks` over `MapHealthChecks` is controlling exactly where health checks runs in the middleware pipeline.
+
+<xref:Microsoft.AspNetCore.Builder.HealthCheckApplicationBuilderExtensions.UseHealthChecks%2A>:
+* Terminates the pipeline when a request matches the health check endpoint. [Short-circuiting](xref:fundamentals/middleware/index) is often desirable because it avoids unnecessary work, such as logging and other middleware.
+* Is primarily used for configuring the health check middleware in the pipeline.
+* Can match any path on a port with a `null` or empty `PathString`, which can be useful for the following scenarios:
+  * Provides a default health check without specifying a path. Allows performing a health check on the root path or any request made to the specified port.
+  * Selecting any endpoint on a port.
+  * Provides a catch-all health check. Handles health checks for any path.
```
Comment (JamesNK) [inline aspnetcore/host-and-deploy/health-checks.md:318]:
```suggestion
* Can match any path on a port with a `null` or empty `PathString`. Allows performing a health check on any request made to the specified port.
```

**Pull request merged**